conn_test.go

```diff
@@ -1190,3 +1275,373 @@ func TestDTLS(t *testing.T) {
+	serverAlert = server.Handshake()
+	assertEquals(t, serverAlert, AlertWouldBlock)
+	assertEquals(t, server.GetHsState(), StateServerStart)
+
+	// Release ClientHello
+	cbConn.Flush()
+
+	// Process ClientHello, send server first flight.
+	states := []State{StateServerNegotiated, StateServerWaitFlight2, StateServerWaitFinished}
+	for _, state := range states {
+		serverAlert = server.Handshake()
+		assertEquals(t, serverAlert, AlertNoAlert)
+		assertEquals(t, server.GetHsState(), state)
+	}
+	serverAlert = server.Handshake()
+	assertEquals(t, serverAlert, AlertWouldBlock)
+
+	clientAlert = client.Handshake()
+	assertEquals(t, clientAlert, AlertWouldBlock)
+
+	// Release server first flight
+	sbConn.Flush()
+	states = []State{StateClientWaitEE, StateClientWaitCertCR, StateClientWaitCV, StateClientWaitFinished, StateClientConnected}
+	for _, state := range states {
+		clientAlert = client.Handshake()
+		assertEquals(t, client.GetHsState(), state)
+		assertEquals(t, clientAlert, AlertNoAlert)
+	}
+
+	serverAlert = server.Handshake()
+	assertEquals(t, serverAlert, AlertWouldBlock)
+	assertEquals(t, server.GetHsState(), StateServerWaitFinished)
+
+	// Release client's second flight.
+	cbConn.Flush()
+	serverAlert = server.Handshake()
+	assertEquals(t, serverAlert, AlertNoAlert)
+	assertEquals(t, server.GetHsState(), StateServerConnected)
+
+	assertDeepEquals(t, client.state.Params, server.state.Params)
+	assertCipherSuiteParamsEquals(t, client.state.cryptoParams, server.state.cryptoParams)
+	assertByteEquals(t, client.state.resumptionSecret, server.state.resumptionSecret)
+	assertByteEquals(t, client.state.clientTrafficSecret, server.state.clientTrafficSecret)
+	assertByteEquals(t, client.state.serverTrafficSecret, server.state.serverTrafficSecret)
+
+	buf := []byte{'a', 'b', 'c'}
+	n, err := client.Write(buf)
+	assertNotError(t, err, "Couldn't write")
+	assertEquals(t, n, len(buf))
+
+	// read := make([]byte, 5)
+	// n, err = server.Read(buf)
+}
+
+func TestTimeoutAndRetransmissionDTLS(t *testing.T) {
+	cConn, sConn := pipe()
+
+	// Wrap these in a buffer so we can simulate blocking
+	cbConn := newBufferedConn(cConn)
+	sbConn := newBufferedConn(sConn)
+
+	client := Client(cbConn, nbConfigDTLS)
+	server := Server(sbConn, nbConfigDTLS)
+
+	var clientAlert, serverAlert Alert
+
+	// Send ClientHello
+	clientAlert = client.Handshake()
+	assertEquals(t, clientAlert, AlertNoAlert)
+	assertEquals(t, client.GetHsState(), StateClientWaitSH)
+	serverAlert = server.Handshake()
+	assertEquals(t, serverAlert, AlertWouldBlock)
+	assertEquals(t, server.GetHsState(), StateServerStart)
+
+	// Simulate loss for the ClientHello
+	cbConn.Clear()
+
+	// Only client should be running a timer.
+	waiting, timeout := server.GetDTLSTimeout()
+	assertTrue(t, !waiting, fmt.Sprintf("Server timer armed: %v", timeout))
+
+	waiting, timeout = client.GetDTLSTimeout()
+	assertTrue(t, waiting, "Client timer not armed")
+
+	// Now check the timer.
+	time.Sleep(timeout)
```

Comment:
You should be careful with this; IIRC go times out tests after a few seconds.

It might be worth considering moving away from using `time.Now()` directly in the timer code.  That would enable you to just advance time forward as you like, so that you don't have to sleep.  Boulder uses https://github.com/jmhodges/clock for this purpose. 

However, synthetic clocks (including the one above) typically don't try to implement a parallel to `time.Timer`, so maybe there's a choice here between using `time.Timer + time.Now() +time.Sleep()`, or rolling our own timer class and using a synthetic clock.

Reply:
Right. I wrote my own timer class so I could do a synthetic clock and then got lazy.

Reply:
In that case, maybe add `TODO(ekr@rtfm.com): Make these independent of the system clock`.